### PR TITLE
(dsl): Support `geoPolygon` query

### PR DESCRIPTION
**Pull request by LeonaNedeljkovic**
Part of #91

### CODE REVIEW

File: modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala [hunk 1]
```diff
@@ -265,6 +265,45 @@ object ElasticQuery {
...
+   * an instance of [[zio.elasticsearch.query.GeoPolygonQuery]] that represents `geo_polygon` query to be performed.
+   */
+  final def geoPolygon[S](field: Field[S, _], coordinates: List[String]): GeoPolygonQuery[S] =
+    GeoPolygon(
+      field = field.toString,
+      points = coordinates,
+      queryName = None,
+      validationMethod = None
+    )
+
```
Comment (markaya) [inline modules/library/src/main/scala/zio/elasticsearch/ElasticQuery.scala:287]: Remove one empty line

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 2]
```diff
@@ -420,6 +420,59 @@ private[elasticsearch] final case class GeoDistance[S](
...
+  def name(value: String): GeoPolygonQuery[S] = self.copy(queryName = Some(value))
+
+  def validationMethod(value: ValidationMethod): GeoPolygonQuery[S] = self.copy(validationMethod = Some(value))
+
+  private[elasticsearch] def toJson(fieldPath: Option[String]): Json =
+    Obj(
+      "geo_polygon" -> Obj(
+        Chunk(
+          Some(field -> Obj("points" -> Arr(points.map(Json.Str(_)): _*))),
+          queryName.map(qn => "_name" -> qn.toJson),
```
Comment (drmarjanovic) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:468]: Let's use `queryName.map("_name" -> _.toJson)`.

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 3]
```diff
@@ -420,6 +420,59 @@ private[elasticsearch] final case class GeoDistance[S](
...
+          validationMethod.map(vm => "validation_method" -> vm.toString.toJson)
```
Comment (drmarjanovic) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:469]: Let's use `validationMethod.map("validation_method" -> _.toString.toJson)`.

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 4]
```diff
@@ -650,6 +650,69 @@ object ElasticQuerySpec extends ZIOSpecDefault {
...
+            equalTo(
+              GeoPolygon[Any](
+                field = "testField",
+                points = List("40, -70", "30, -80", "20, -90"),
+                queryName = None,
+                validationMethod = None
+              )
+            )
+          ) &&
+          assert(queryString)(
```
Comment (drmarjanovic) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:678]: Should we move this in the row above?

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 5]
```diff
@@ -2019,6 +2082,88 @@ object ElasticQuerySpec extends ZIOSpecDefault {
...
+            geoPolygon(TestDocument.stringField, List("40, -70", "30, -80", "20, -90")).name("name")
+          val queryWithValidationMethod =
+            geoPolygon(TestDocument.stringField, List("40, -70", "30, -80", "20, -90")).validationMethod(
+              IgnoreMalformed
+            )
+          val queryWithAllParams = geoPolygon(TestDocument.stringField, List("40, -70", "30, -80", "20, -90"))
+            .validationMethod(IgnoreMalformed)
+            .name("name")
+
+          println(queryWithName.toJson(fieldPath = None))
```
Comment (drmarjanovic) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:2100]: Please remove `println`.

**Review by drmarjanovic — CHANGES_REQUESTED** [review]

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 6]
```diff
@@ -420,6 +420,59 @@ private[elasticsearch] final case class GeoDistance[S](
...
 }
 
+sealed trait GeoPolygonQuery[S] extends ElasticQuery[S] {
+
+  /**
+   * Sets the `queryName` parameter for the [[zio.elasticsearch.query.GeoPolygonQuery]]. Represents the optional name
+   * field to identify the query
+   *
+   * @param value
+   *   the [[String]] value to represent the name field
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:430]:
```suggestion
   *   the text value to represent the name field
```

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 7]
```diff
@@ -420,6 +420,59 @@ private[elasticsearch] final case class GeoDistance[S](
...
+}
+
+private[elasticsearch] final case class GeoPolygon[S](
+  field: String,
+  points: List[String],
+  queryName: Option[String],
+  validationMethod: Option[ValidationMethod]
+) extends GeoPolygonQuery[S] { self =>
+
+  def name(value: String): GeoPolygonQuery[S] = self.copy(queryName = Some(value))
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:459]: Move `self.copy(queryName = Some(value))` to new line.

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

**Review by markaya — DISMISSED** [review]

**Review by dbulaja98 — APPROVED** [review]

File: modules/library/src/it/scala/zio/elasticsearch/HttpExecutorSpec.scala [hunk 8]
```diff
@@ -1706,6 +1706,41 @@ object HttpExecutorSpec extends IntegrationSpec {
...
+                r1 <- Executor
+                        .execute(
+                          ElasticRequest.search(
+                            geoPolygonIndex,
+                            ElasticQuery
+                              .geoPolygon("locationField", List("0, 0", "0, 90", "90, 90", "90, 0"))
+                          )
+                        )
+                        .documentAs[TestDocument]
+              } yield assertTrue(r1 == Chunk(document))
```
Comment (markaya) [inline modules/library/src/it/scala/zio/elasticsearch/HttpExecutorSpec.scala:1740]:
```suggestion
              } yield assert(equalTo(Chunk(document)))
```

**Review by drmarjanovic — APPROVED** [review]

**Review by dbulaja98 — DISMISSED** [review]

**Comment by drmarjanovic** [comment]
@LeonaNedeljkovic, are you still working on this? Is there anything that's blocking you?

File: modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala [hunk 9]
```diff
@@ -534,6 +534,61 @@ private[elasticsearch] final case class GeoDistance[S](
...
+   *   - [[zio.elasticsearch.query.ValidationMethod.Coerce]]: Additionally try and infer correct coordinates
+   * @return
+   *   an instance of [[zio.elasticsearch.query.GeoPolygonQuery]] enriched with the `validationMethod` parameter.
+   */
+  def validationMethod(value: ValidationMethod): GeoPolygonQuery[S]
+}
+
+private[elasticsearch] final case class GeoPolygon[S](
+  field: String,
+  points: List[String],
```
Comment (dbulaja98) [inline modules/library/src/main/scala/zio/elasticsearch/query/Queries.scala:568]: Can you replace `List[String]` with `Chunk[String]`?

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 10]
```diff
@@ -813,6 +813,68 @@ object ElasticQuerySpec extends ZIOSpecDefault {
             )
           )
         },
+        test("geoPolygon") {
+          val query =
+            geoPolygon("testField", List("40, -70", "30, -80", "20, -90"))
+          val queryString =
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:819]:
```suggestion
          val queryTs =
```

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 11]
```diff
@@ -2910,6 +2972,87 @@ object ElasticQuerySpec extends ZIOSpecDefault {
           assert(queryWithValidationMethod.toJson(fieldPath = None))(equalTo(expectedWithValidationMethod.toJson)) &&
           assert(queryWithAllParams.toJson(fieldPath = None))(equalTo(expectedWithAllParams.toJson))
         },
+        test("geoPolygon") {
+          val query =
+            geoPolygon("testField", List("40, -70", "30, -80", "20, -90"))
+          val queryString =
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:2978]:
```suggestion
          val queryTs =
```

File: modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala [hunk 12]
```diff
@@ -2910,6 +2972,87 @@ object ElasticQuerySpec extends ZIOSpecDefault {
...
+              |{
+              |  "geo_polygon": {
+              |    "testField": {
+              |       "points": ["40, -70", "30, -80", "20, -90"]
+              |    }
+              |  }
+              |}
+              |""".stripMargin
+
+          val expectedWithString =
```
Comment (dbulaja98) [inline modules/library/src/test/scala/zio/elasticsearch/ElasticQuerySpec.scala:3002]:
```suggestion
          val expectedTs =
```

File: docs/overview/queries/elastic_query_geo_polygon.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+```scala
+val query: GeoPolygonQuery = geoPolygon(field = "location", List("0, 0", "0, 90", "90, 90", "90, 0"))
+```
+
+You can create a [type-safe](https://lambdaworks.github.io/zio-elasticsearch/overview/overview_zio_prelude_schema) `GeoPolygon` query using the `geoPolygon` method with list of coordinates in the following manner:
+```scala
+val query: GeoPolygonQuery = geoPolygon(field = Document.location, List("0, 0", "0, 90", "90, 90", "90, 0"))
+```
+
+
```
Comment (dbulaja98) [inline docs/overview/queries/elastic_query_geo_polygon.md:24]: Omit one empty line.

**Review by dbulaja98 — CHANGES_REQUESTED** [review]

**Review by dbulaja98 — APPROVED** [review]

**Pull request merged**